autotest.pm

```diff
@@ -345,6 +350,44 @@ sub loadtestdir {
     }
 }
 
+sub parse_serial_output {
+    return -1 if !-e 'serial0';
+
+    my ($serial_file_pos, $t) = @_;
+    my $failures = $t->{serial_failures};
+
+    open my $ifh, '<', 'serial0' or die "Unable to open file: $!";
+    seek $ifh, $serial_file_pos, 0;
+    my $die = 0;
+    while (defined(my $line = <$ifh>)) {
+        chomp $line;
+        for my $type (keys %$failures) {
+            for my $regexp (@{$failures->{$type}}) {
+                my $re = quotemeta $regexp;
+                if ($line =~ /$regexp/ || $line =~ /$re/) {
+                    if ($type eq 'soft') {
+                        $t->record_testresult('softfail');
+                        $t->record_resultfile('Serial Failure', "Serial error: $line", result => 'softfail');
+                        $t->{result} = 'softfail';
+
+                    }
+                    else {
```

Comment:
if you are only allowing `soft` or `hard`, this block could be simplified, since you can already map:
hard = fail && $die = 1;
soft = softfail && $die = 0;